File: cpp/04_009.cpp
#include <string>
#include <memory>
#include <stdexcept>

class User; // Forward declaration
class UserRepository; // Forward declaration
class UserAlreadyExistsException; // Forward declaration

class UserService {
private:
    std::shared_ptr<UserRepository> userRepository;
    
    bool userAlreadyExists(const std::string& username) {
        return userRepository->existsByUsername(username);
    }
    
public:
    UserService(std::shared_ptr<UserRepository> repository) 
        : userRepository(repository) {}
 
    User getUser(const std::string& username) {
        return userRepository->findByUsername(username);
    }
 
    void createUser(const User& user) {
        if (userAlreadyExists(user.getUsername())) {
            throw UserAlreadyExistsException();
        }
        userRepository->save(user);
    }
};

File: data structures and algorithms/cs/04_009.cs
class UserService
{
   private readonly UserRepository userRepository;

   public UserService(UserRepository userRepository)
   {
      this.userRepository = userRepository;
   }

   public User GetUser(string username)
   {
      return userRepository.FindByUsername(username);
   }

   public void CreateUser(User user)
   {
      if (UserAlreadyExists(user.Username))
      {
         throw new UserAlreadyExistsException();
      }
      userRepository.Save(user);
   }

   private bool UserAlreadyExists(string username)
   {
      return userRepository.ExistsByUsername(username);
   }
}
